component/dialer/sockopt_windows.go

```diff
@@ -0,0 +1,67 @@
+package dialer
+
+import (
+	"encoding/binary"
+	"net"
+	"syscall"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+)
+
+func setSocketOptions(network, address string, c syscall.RawConn, opts *Options) (err error) {
+	if opts == nil || !isTCPSocket(network) && !isUDPSocket(network) {
+		return
+	}
+
+	var innerErr error
+	err = c.Control(func(fd uintptr) {
+		host, _, _ := net.SplitHostPort(address)
+		if ip := net.ParseIP(host); ip != nil && !ip.IsGlobalUnicast() {
+			return
+		}
+
+		if opts.InterfaceIndex == 0 && opts.InterfaceName != "" {
+			if iface, err := net.InterfaceByName(opts.InterfaceName); err == nil {
+				opts.InterfaceIndex = iface.Index
+			}
+		}
+
+		if opts.InterfaceIndex != 0 {
+			switch network {
+			case "tcp4", "udp4":
+				innerErr = bindSocketToInterface4(windows.Handle(fd), uint32(opts.InterfaceIndex))
+			case "tcp6", "udp6":
+				innerErr = bindSocketToInterface6(windows.Handle(fd), uint32(opts.InterfaceIndex))
+				if network == "udp6" {
+					// the underlying IP net maybe IPv4 even if the 'network' param is 'udp6',
+					// so we should bind socket to interface4 at the same time
+					innerErr = bindSocketToInterface4(windows.Handle(fd), uint32(opts.InterfaceIndex))
+				}
+			}
+		}
+	})
+
+	if innerErr != nil {
+		err = innerErr
+	}
+	return
+}
+
+func bindSocketToInterface4(handle windows.Handle, interfaceIndex uint32) error {
+	const IP_UNICAST_IF = 31
+	/* MSDN says for IPv4 this needs to be in net byte order, so that it's like an IP address with leading zeros. */
+	var bytes [4]byte
+	binary.BigEndian.PutUint32(bytes[:], interfaceIndex)
+	interfaceIndex = *(*uint32)(unsafe.Pointer(&bytes[0]))
```

Comment:
So, if I get this correctly, the whole point here is to convert the index uint32 into big endian uint32?

I think this can be done via bit operations.

Reply:
For example, in Python:

```python
def reverse_int(x):
    x = ((x & 0x55555555) << 1) | ((x & 0xAAAAAAAA) >> 1)
    x = ((x & 0x33333333) << 2) | ((x & 0xCCCCCCCC) >> 2)
    x = ((x & 0x0F0F0F0F) << 4) | ((x & 0xF0F0F0F0) >> 4)
    x = ((x & 0x00FF00FF) << 8) | ((x & 0xFF00FF00) >> 8)
    x = ((x & 0x0000FFFF) << 16) | ((x & 0xFFFF0000) >> 16)
    return x
```

Reply:
I don't understand the code above. And I try to transfer it to go,  like below:
```go
func ReverseInt(x uint32) uint32 {
	x = ((x & 0x55555555) << 1) | ((x & 0xAAAAAAAA) >> 1)
	x = ((x & 0x33333333) << 2) | ((x & 0xCCCCCCCC) >> 2)
	x = ((x & 0x0F0F0F0F) << 4) | ((x & 0xF0F0F0F0) >> 4)
	x = ((x & 0x00FF00FF) << 8) | ((x & 0xFF00FF00) >> 8)
	x = ((x & 0x0000FFFF) << 16) | ((x & 0xFFFF0000) >> 16)
	return x
}
```
I tested in my local, But it can't get the right answer. Could you give me more help here?

Reply:
Oh sorry, I misunderstood the logic here.

Reply:
So the value of index is determined by the prossor's endian? Do you have the specific link explaining for it?

Reply:
There are some helpful references that I found:
1. https://stackoverflow.com/questions/51332658/any-better-way-to-check-endianness-in-go

2. https://stackoverflow.com/questions/7184789/does-bit-shift-depend-on-endianness

3. https://segmentfault.com/a/1190000039738719

Reply:
Sorry, I mean why would you do this conversion for windows syscall?

Reply:
Oh, yes. This link explains why we should do that: https://learn.microsoft.com/en-us/windows/win32/winsock/ipproto-ip-socket-options

Please search "IP_UNICAST_IF" in this page, and see the description. 
```
This DWORD parameter must be an interface index in network byte order.
```